erpnext/manufacturing/doctype/bom_update_log/bom_update_log.py

```diff
@@ -93,6 +93,7 @@ def on_submit(self):
 		else:
 			frappe.enqueue(
 				method="erpnext.manufacturing.doctype.bom_update_log.bom_update_log.process_boms_cost_level_wise",
+				queue="long",
```

Comment:
The first query itself is timing out, will increasing the timeout help much here? 

Is this query optimized? `bom_no` index and removal of infull might help. 


https://github.com/frappe/erpnext/blob/3bc79eebe38a568b87cbb11b50de766dd6c23ed8/erpnext/manufacturing/doctype/bom_update_log/bom_updation_utils.py#L157-L165

Reply:
kinda got it wrong, set diff is all parent BOMs.... but still removing ifnull/trying join might give better results here.

Reply:
Btw ran the query, it only took 2 sec. So it's not even the main problem. 

:detective:

Reply:
```168202065 function calls (164446933 primitive calls) in 260.553 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    522/1    0.001    0.000  260.553  260.553 {built-in method builtins.exec}
        1    0.000    0.000  260.553  260.553 bom_update_tool.py:33(enqueue_update_cost)
        1    0.000    0.000  260.553  260.553 bom_update_tool.py:52(create_bom_update_log)
        1    0.000    0.000  260.359  260.359 document.py:1017(submit)
        1    0.000    0.000  260.359  260.359 document.py:1005(_submit)
        1    0.000    0.000  260.359  260.359 document.py:310(save)
        1    0.000    0.000  260.359  260.359 document.py:314(_save)
        1    0.000    0.000  260.359  260.359 document.py:221(insert)
    14/10    0.000    0.000  260.248   26.025 document.py:931(run_method)
    14/10    0.000    0.000  260.221   26.022 document.py:1252(composer)
    14/10    0.000    0.000  260.210   26.021 document.py:1243(runner)
        1    0.000    0.000  260.204  260.204 document.py:1076(run_post_save_methods)
    14/10    0.000    0.000  260.177   26.018 document.py:934(fn)
        1    0.000    0.000  260.168  260.168 bom_update_log.py:69(on_submit)
      3/1    0.000    0.000  260.168  260.168 __init__.py:1992(enqueue)
      3/1    0.000    0.000  260.168  260.168 background_jobs.py:47(enqueue)
      3/1    0.000    0.000  260.168  260.168 __init__.py:1467(call)
        1    0.000    0.000  260.168  260.168 bom_update_log.py:111(process_boms_cost_level_wise)
   116637    0.963    0.000  229.235    0.002 database.py:111(sql)
   116637    0.175    0.000  221.985    0.002 cursors.py:129(execute)
   116637    0.188    0.000  217.539    0.002 cursors.py:306(_query)
   116637    0.240    0.000  217.118    0.002 connections.py:542(query)
   116637    0.172    0.000  213.787    0.002 connections.py:763(_read_query_result)
   116637    0.219    0.000  213.518    0.002 connections.py:1154(read)
  1610400    4.264    0.000  200.112    0.000 connections.py:683(_read_packet)
  3220800    2.550    0.000  178.181    0.000 connections.py:728(_read_bytes)
  3220802    1.613    0.000  173.482    0.000 {method 'read' of '_io.BufferedReader' objects}
   117722    0.228    0.000  171.869    0.001 socket.py:575(readinto)
   132725  171.794    0.001  171.794    0.001 {method 'recv_into' of '_socket.socket' objects}
    88511    0.145    0.000  153.593    0.002 connections.py:1233(_read_result_packet)
        1    0.000    0.000  141.080  141.080 bom_updation_utils.py:157(get_leaf_boms)
        1    0.000    0.000  141.080  141.080 database.py:286(sql_list)
    88511    0.321    0.000  123.434    0.001 connections.py:1266(_read_rowdata_packet)
        1    0.000    0.000  119.087  119.087 bom_update_log.py:146(queue_bom_cost_jobs)
        2    0.000    0.000  119.071   59.536 bom_updation_utils.py:46(update_cost_in_level)
        2    0.036    0.018  119.049   59.524 bom_updation_utils.py:117(update_cost_in_boms)
    62956    0.115    0.000   61.401    0.001 database.py:487(get_values)
    62956    0.262    0.000   61.271    0.001 database.py:708(_get_values_from_table)
     7265    0.048    0.000   57.574    0.008 __init__.py:1062(get_doc)
     7106    0.032    0.000   56.780    0.008 bom.py:592(calculate_cost)
     7265    0.027    0.000   56.453    0.008 document.py:30(get_doc)
53519/7274    0.182    0.000   56.260    0.008 document.py:87(__init__)
7264/7263    0.184    0.000   56.034    0.008 document.py:144(load_from_db)
     7106    0.017    0.000   54.994    0.008 website_generator.py:17(__init__)
     7106    0.233    0.000   51.056    0.007 bom.py:647(calculate_rm_cost)
    22455    0.116    0.000   39.940    0.002 bom.py:351(get_rm_rate)
    22455    0.054    0.000   32.120    0.001 bom.py:911(get_bom_item_rate)
    21667    0.825    0.000   30.561    0.001 bom.py:946(get_valuation_rate)
    88511    2.606    0.000   29.925    0.000 connections.py:1298(_get_descriptions)
    33764    0.053    0.000   22.753    0.001 database.py:435(get_value)
    25443    0.075    0.000   20.626    0.001 utils.py:64(execute_query)
    28119    0.257    0.000   18.243    0.001 base_document.py:447(db_update)
  1206121    0.703    0.000   16.832    0.000 protocol.py:234(__init__)
  1206121    3.200    0.000   15.868    0.000 protocol.py:238(_parse_field_descriptor)
  9040508    3.798    0.000   12.959    0.000 protocol.py:165(read_length_coded_string)
   110616    1.387    0.000    6.926    0.000 connections.py:1279(_read_row_from_packet)
    25443    0.102    0.000    5.591    0.000 utils.py:68(prepare_query)
     7106    0.062    0.000    5.512    0.001 bom.py:703(calculate_exploded_cost)
    25443    0.076    0.000    5.333    0.000 dialects.py:129(get_sql)
    25443    0.284    0.000    5.194    0.000 queries.py:1218(get_sql)
  9185275    2.857    0.000    5.016    0.000 protocol.py:147(read_length_encoded_integer)
    28574    2.252    0.000    4.757    0.000 base_document.py:274(get_valid_dict)
  8731899    3.744    0.000    4.276    0.000 protocol.py:62(read)
   178093    0.260    0.000    4.212    0.000 utils.py:48(_copy)
   116637    0.314    0.000    4.087    0.000 cursors.py:115(mogrify)
   116618    0.316    0.000    3.754    0.000 cursors.py:105(_escape_args)
    89826    0.343    0.000    3.584    0.000 base_document.py:954(precision)
309258/200775    0.199    0.000    3.320    0.000 {method 'join' of 'str' objects}
   116639    0.808    0.000    3.019    0.000 connections.py:787(_execute_command)
   865983    0.540    0.000    2.979    0.000 local.py:345(__getattr__)```

Reply:
```
171906 function calls (171733 primitive calls) in 83.689 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     24/1    0.000    0.000   83.689   83.689 {built-in method builtins.exec}
        1    0.000    0.000   83.689   83.689 <string>:1(<module>)
        1    0.000    0.000   83.689   83.689 bom_updation_utils.py:157(get_leaf_boms)
        1    0.000    0.000   83.689   83.689 database.py:286(sql_list)
        1    0.000    0.000   83.689   83.689 database.py:111(sql)
        1    0.000    0.000   83.678   83.678 cursors.py:129(execute)
        1    0.000    0.000   83.678   83.678 cursors.py:306(_query)
        1    0.000    0.000   83.678   83.678 connections.py:542(query)
        1    0.000    0.000   83.678   83.678 connections.py:763(_read_query_result)
        1    0.000    0.000   83.678   83.678 connections.py:1154(read)
     7117    0.016    0.000   83.646    0.012 connections.py:683(_read_packet)
    14234    0.010    0.000   83.626    0.006 connections.py:728(_read_bytes)
    14234    0.005    0.000   83.607    0.006 {method 'read' of '_io.BufferedReader' objects}
       38    0.000    0.000   83.602    2.200 socket.py:575(readinto)
       38   83.602    2.200   83.602    2.200 {method 'recv_into' of '_socket.socket' objects}
        1    0.000    0.000   73.674   73.674 connections.py:1233(_read_result_packet)
        1    0.007    0.007   73.674   73.674 connections.py:1266(_read_rowdata_packet)
     7109    0.007    0.000    0.021    0.000 connections.py:1279(_read_row_from_packet)
     7115    0.003    0.000    0.012    0.000 protocol.py:165(read_length_coded_string)
        1    0.000    0.000    0.010    0.010 database.py:82(connect)
        1    0.000    0.000    0.009    0.009 database.py:1288(get_query_execution_timeout)
     27/1    0.000    0.000    0.009    0.009 <frozen importlib._bootstrap>:978(_find_and_load)
     27/1    0.000    0.000    0.009    0.009 <frozen importlib._bootstrap>:948(_find_and_load_unlocked)
     24/1    0.000    0.000    0.008    0.008 <frozen importlib._bootstrap>:663(_load_unlocked)
     23/1    0.000    0.000    0.008    0.008 <frozen importlib._bootstrap_external>:722(exec_module)
     30/1    0.000    0.000    0.008    0.008 <frozen importlib._bootstrap>:211(_call_with_frames_removed)
        1    0.000    0.000    0.008    0.008 __init__.py:3(<module>)
    14240    0.008    0.000    0.008    0.000 {method 'settimeout' of '_socket.socket' objects}
     7120    0.003    0.000    0.004    0.000 protocol.py:147(read_length_encoded_integer)
        1    0.000    0.000    0.004    0.004 worker.py:2(<module>)
     7115    0.003    0.000    0.004    0.000 protocol.py:62(read)
     7110    0.002    0.000    0.003    0.000 connections.py:1221(_check_packet_is_eof)
       23    0.000    0.000    0.003    0.000 <frozen importlib._bootstrap_external>:793(get_code)
       27    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap>:882(_find_spec)
     7120    0.002    0.000    0.002    0.000 protocol.py:114(read_uint8)
       26    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap_external>:1272(find_spec)
       26    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap_external>:1240(_get_spec)
       23    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap_external>:523(_compile_bytecode)
       23    0.002    0.000    0.002    0.000 {built-in method marshal.loads}
        1    0.000    0.000    0.002    0.002 scheduler.py:1(<module>)
       60    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap_external>:1356(find_spec)
     7115    0.002    0.000    0.002    0.000 {method 'decode' of 'bytes' objects}
        1    0.000    0.000    0.002    0.002 job.py:2(<module>)
     7120    0.002    0.000    0.002    0.000 {built-in method _struct.unpack}
   106/99    0.000    0.000    0.002    0.000 <frozen importlib._bootstrap>:1009(_handle_fromlist)
21557/21551    0.001    0.000    0.001    0.000 {built-in method builtins.len}
       59    0.001    0.000    0.001    0.000 {built-in method builtins.__build_class__}
        1    0.000    0.000    0.001    0.001 database.py:58(get_connection)
      5/2    0.000    0.000    0.001    0.001 {built-in method builtins.__import__}
```